relayer/relayer.go

```diff
@@ -0,0 +1,157 @@
+)
+
+var commitEvent = abi.MustNewEvent(`event NewBundleCommit(uint256 startId, uint256 endId, bytes32 root)`)
+
+type Relayer struct {
+	dataDir           string
+	rpcEndpoint       string
+	stateReceiverAddr ethgo.Address
+	logger            hcf.Logger
+	client            *jsonrpc.Client
+	txRelayer         txrelayer.TxRelayer
+	key               ethgo.Key
+}
+
+func sanitizeRPCEndpoint(rpcEndpoint string) string {
+	if rpcEndpoint == "" || strings.Contains(rpcEndpoint, "0.0.0.0") {
+		_, port, err := net.SplitHostPort(rpcEndpoint)
+		if err == nil {
+			rpcEndpoint = fmt.Sprintf("http://%s:%s", "localhost", port)
+		} else {
+			rpcEndpoint = "http://localhost:8545"
+		}
+	}
+
+	return rpcEndpoint
+}
+
+func NewRelayer(
+	dataDir string,
+	rpcEndpoint string,
+	stateReceiverAddr ethgo.Address,
+	logger hcf.Logger,
+	key ethgo.Key,
+) *Relayer {
+	endpoint := sanitizeRPCEndpoint(rpcEndpoint)
+
+	// create the JSON RPC client
+	client, err := jsonrpc.NewClient(endpoint)
+	if err != nil {
+		logger.Error("Failed to create the JSON RPC client", "err", err)
+
+		return nil
+	}
+
+	txRelayer, err := txrelayer.NewTxRelayer(txrelayer.WithClient(client))
+
+	return &Relayer{
+		dataDir:           dataDir,
+		rpcEndpoint:       endpoint,
+		stateReceiverAddr: stateReceiverAddr,
+		logger:            logger,
+		client:            client,
+		txRelayer:         txRelayer,
+		key:               key,
+	}
+}
+
+func (r *Relayer) Start() error {
+	et := eventTracker{
+		dataDir:           r.dataDir,
+		rpcEndpoint:       r.rpcEndpoint,
+		stateReceiverAddr: r.stateReceiverAddr,
+		subscriber:        r,
+		logger:            r.logger,
+	}
+
+	return et.start()
+}
+
+func (r *Relayer) AddLog(log *ethgo.Log) {
+	r.logger.Info("Received a log", "log", log)
+
+	if commitEvent.Match(log) {
+		vals, err := commitEvent.ParseLog(log)
+		if err != nil {
+			panic(err)
+		}
+
+		var startID uint64
+		if sid, ok := vals["startId"].(big.Int); ok {
+			startID = sid.Uint64()
+		}
+
+		var endID uint64
+		if eid, ok := vals["endId"].(big.Int); ok {
+			endID = eid.Uint64()
+		}
+
+		r.logger.Info("Commit: Block %d StartID %d EndID %d\n", log.BlockNumber, startID, endID)
+
+		for i := startID; i < endID; i++ {
+			// query the state sync proof
+			stateSyncProof, err := r.queryStateSyncProof(strconv.Itoa(int(i)))
+			if err != nil {
+				r.logger.Error("Failed to query state sync proof", "err", err)
+			}
+
+			if err := r.executeStateSync(stateSyncProof); err != nil {
+				r.logger.Error("Failed to execute state sync", "err", err)
+			}
+		}
+	}
+}
+
+// queryStateSyncProof queries the state sync proof
+func (r *Relayer) queryStateSyncProof(stateSyncID string) (*types.StateSyncProof, error) {
+	// retrieve state sync proof
+	var stateSyncProof types.StateSyncProof
+
+	err := r.client.Call("bridge_getStateSyncProof", &stateSyncProof, stateSyncID)
+	if err != nil {
+		return nil, err
+	}
+
+	r.logger.Debug("state sync proof:", stateSyncProof)
```

Comment:
Logger expects odd number of arguments:

```suggestion
	r.logger.Debug(fmt.Sprintf("state sync proof: %v", stateSyncProof))
```